File: backend/app/schemas/post_likes.py
from uuid import UUID

from pydantic import BaseModel, Field


class PostLikeTarget(BaseModel):
    target_character_id: UUID
    post_id: str = Field(min_length=1, max_length=120)


class PostLikesQuery(BaseModel):
    liker_account_id: str = Field(min_length=1, max_length=120)
    targets: list[PostLikeTarget] = Field(default_factory=list, max_length=100)


class PostLikeUpdate(PostLikeTarget):
    liker_account_id: str = Field(min_length=1, max_length=120)
    liked: bool


class PostLikeItem(PostLikeTarget):
    available: bool
    liked: bool
    likes: int = Field(ge=0)


class PostLikesResponse(BaseModel):
    items: list[PostLikeItem] = Field(default_factory=list)

File: apps/frontend/src/api/postLikes.ts
import { apiJson } from "./client.js";


export type PostLikeTarget = {
  target_shared_character_id: string;
  post_id: string;
};

export type PostLikeItem = PostLikeTarget & {
  available: boolean;
  liked: boolean;
  likes: number;
};

type PostLikesResponse = {
  items: PostLikeItem[];
};

export async function queryPostLikes(likerAccountId: string, targets: PostLikeTarget[]): Promise<PostLikeItem[]> {
  const batches = chunkTargets(targets);
  const responses = await Promise.all(batches.map((batch) => queryPostLikeBatch(likerAccountId, batch)));
  return responses.flatMap((response) => response.items);
}

async function queryPostLikeBatch(likerAccountId: string, targets: PostLikeTarget[]): Promise<PostLikesResponse> {
  const response = await apiJson<PostLikesResponse>("/post-likes/query", {
    method: "POST",
    body: JSON.stringify({ liker_account_id: likerAccountId, targets }),
  });
  return response;
}

export function updatePostLike(likerAccountId: string, target: PostLikeTarget, liked: boolean): Promise<PostLikeItem> {
  return apiJson<PostLikeItem>("/post-likes", {
    method: "PUT",
    body: JSON.stringify({ liker_account_id: likerAccountId, ...target, liked }),
  });
}

function chunkTargets(targets: PostLikeTarget[]): PostLikeTarget[][] {
  const chunks: PostLikeTarget[][] = [];
  for (let index = 0; index < targets.length; index += 100) chunks.push(targets.slice(index, index + 100));
  return chunks;
}
